src/librustdoc/clean/mod.rs

```diff
@@ -1877,35 +1821,29 @@ impl Clean<VariantStruct> for rustc_hir::VariantData<'_> {
 
 impl Clean<Item> for doctree::Enum<'_> {
     fn clean(&self, cx: &DocContext<'_>) -> Item {
-        Item {
-            name: Some(self.name.clean(cx)),
-            attrs: self.attrs.clean(cx),
-            source: self.span.clean(cx),
-            def_id: cx.tcx.hir().local_def_id(self.id).to_def_id(),
-            visibility: self.vis.clean(cx),
-            stability: cx.stability(self.id),
-            deprecation: cx.deprecation(self.id).clean(cx),
-            inner: EnumItem(Enum {
+        Item::from_hir_id_and_parts(
+            self.id,
+            Some(self.name),
+            EnumItem(Enum {
                 variants: self.variants.iter().map(|v| v.clean(cx)).collect(),
                 generics: self.generics.clean(cx),
                 variants_stripped: false,
             }),
-        }
+            cx,
+        )
     }
 }
 
 impl Clean<Item> for doctree::Variant<'_> {
     fn clean(&self, cx: &DocContext<'_>) -> Item {
-        Item {
-            name: Some(self.name.clean(cx)),
-            attrs: self.attrs.clean(cx),
-            source: self.span.clean(cx),
-            visibility: Inherited,
-            stability: cx.stability(self.id),
-            deprecation: cx.deprecation(self.id).clean(cx),
-            def_id: cx.tcx.hir().local_def_id(self.id).to_def_id(),
-            inner: VariantItem(Variant { kind: self.def.clean(cx) }),
-        }
+        let what_rustc_thinks = Item::from_hir_id_and_parts(
+            self.id,
+            Some(self.name),
+            VariantItem(Variant { kind: self.def.clean(cx) }),
+            cx,
+        );
+        // don't show `pub` for variants, which are always public
```

Comment:
This should probably go into rendering of `VariantItem` instead.

Reply:
I would prefer not to do that, at least not in this PR. Right now the rendering is handled here: https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/librustdoc/html/render/mod.rs#L3114

`render_struct` takes a `clean::Item` so it can be used for both variant structs and regular struct items: https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/librustdoc/html/render/mod.rs#L3251-L3259

If I changed this to have the 'real' visibility I'd have to add more state to `render` to tell the difference between a variant and a struct. And I'm not sure this is the only place `VariantItem` is used, there may be others depending on this behavior.

Reply:
Actually it looks like `structhead` is the state I need already. But I'm still not sure this is the only place `VariantItem` is used.

Reply:
Ok.